src/brandedPath.ts

```diff
@@ -0,0 +1,101 @@
+import osPath from "node:path";
+import posixPath from "node:path/posix";
+
+// including "unknown &" here improves type error message
+type BrandedString<T> = unknown &
+  Omit<string, "replace" | "slice"> & {__type: T} & {
+    replace: (search: string | RegExp, replace: string) => BrandedString<T>;
+    slice: (start?: number, end?: number) => BrandedString<T>;
+  };
+
+export type FilePath = BrandedString<"FilePath">;
+export type UrlPath = BrandedString<"UrlPath">;
+
+export function FilePath(path: string | FilePath): FilePath {
+  if (osPath.sep === "\\") {
+    path = path.replaceAll("/", osPath.sep);
+  } else if (osPath.sep === "/") {
+    path = path.replaceAll("\\", osPath.sep);
+  }
+  return path as unknown as FilePath;
+}
+
+export function UrlPath(path: string | UrlPath): UrlPath {
+  path = path.replaceAll("\\", posixPath.sep);
+  return path as unknown as UrlPath;
+}
+
+export function unFilePath(path: FilePath | string): string {
+  return path as unknown as string;
+}
+
+export function unUrlPath(path: UrlPath | string): string {
+  return path as unknown as string;
+}
+
```

Comment:
One thing that I was disappointed about was that I couldn't figure out how to safely tell TypeScript that `UrlPath` and `FilePath` were actually just type safe strings. If we remove the overrides of `replace` and similar in the definitions of the path types, it can understand it, but then we erase types when working on the path.

I think if we wanted to eliminate these functions, we'd have to inline the `as unknown as string` text everywhere we call this function. I chose not to do that because I treat `as`-casts as dangerous code. I'd prefer to isolate it to as few places as possible. Then those isolated places can be thoroughly vetted, either through tests or through reading the code. In this case the cast is trivial to verify, so just having this function is enough.

I agree though that I would prefer some better way to do this, like maybe a "type level function". I don't know of one though.

Reply:
All of the wrapping worries me, but I can't express why. It feels excessive or invasive maybe? I'm definitely wishing we didn't have to do it.

Reply:
This pattern, of having a new type that is just an existing type with a new name is something that I've used in other languages. In those languages there are often nice features built into the language that help make this easier, like implicit conversion, operator overloading, and monad-style-map. TypeScript doesn't really give us any of those things, so it is all a bit clunky.

I could imagine taking out most of these wrappers now (except for the places where we need slash conversion). I don't know how we would prevent new changes to the code base from getting it wrong in the future. Maybe just test coverage and Windows in CI? That feels much less robust than TypeScript to me.